components/forms/src/api.rs

```diff
@@ -0,0 +1,248 @@
+    address.time_last_used = 0;
+    address.time_last_modified = now_ms;
+    address.times_used = 0;
+
+    tx.execute_named_cached(
+        "INSERT OR IGNORE INTO addresses_data (
+            guid,
+            given_name,
+            additional_name,
+            family_name,
+            organization,
+            street_address,
+            address_level3,
+            address_level2,
+            address_level1,
+            postal_code,
+            country,
+            tel,
+            email,
+            time_created,
+            time_last_used,
+            time_last_modified,
+            times_used,
+            sync_change_counter
+        ) VALUES (
+            :guid,
+            :given_name,
+            :additional_name,
+            :family_name,
+            :organization,
+            :street_address,
+            :address_level3,
+            :address_level2,
+            :address_level1,
+            :postal_code,
+            :country,
+            :tel,
+            :email,
+            :time_created,
+            :time_last_used,
+            :time_last_modified,
+            :times_used,
+            1
+        )",
+        rusqlite::named_params! {
+            ":guid": address.guid,
+            ":given_name": address.given_name,
+            ":additional_name": address.additional_name,
+            ":family_name": address.family_name,
+            ":organization": address.organization,
+            ":street_address": address.street_address,
+            ":address_level3": address.address_level3,
+            ":address_level2": address.address_level2,
+            ":address_level1": address.address_level1,
+            ":postal_code": address.postal_code,
+            ":country": address.country,
+            ":tel": address.tel,
+            ":email": address.email,
+            ":time_created": address.time_created,
+            ":time_last_used": address.time_last_used,
+            ":time_last_modified": address.time_last_modified,
+            ":times_used": address.times_used,
+        },
+    )?;
+    Ok(address)
+}
+
+fn get_address(conn: &Connection, guid: &Guid) -> Result<Option<Address>> {
+    let sql = "
+        SELECT
+            guid,
+            given_name,
+            additional_name,
+            family_name,
+            organization,
+            street_address,
+            address_level3,
+            address_level2,
+            address_level1,
+            postal_code,
+            country,
+            tel,
+            email,
+            time_created,
+            time_last_used,
+            time_last_modified,
+            times_used
+        FROM addresses_data
+        WHERE guid = :guid";
+    let address = conn.query_row(
+        sql,
+        &[guid.as_str()],
+        |row| {
+            Ok(Address {
+                guid: Guid::from_string(row.get::<_, String>("guid")?),
+                given_name: row.get::<_, String>("given_name")?,
```

Comment:
```suggestion
                given_name: row.get("given_name")?,
```

I’m not sure if this would compile (and please ignore me if you tried this already, and I’m talking nonsense!), but the type checker might be smart enough to infer this is a `String` without having to write out the turbofish.

Also, holy moly, so many fields! 😅

Reply:
It compiles! Love it, so much cleaner.